modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/AndesMessageMetadata.java

```diff
@@ -315,15 +337,40 @@ private void parseMetaData() {
         buf = buf.slice();
         MessageMetaDataType type = MessageMetaDataType.values()[originalMetadata[0]];
         metaDataType = type;
-        StorableMessageMetaData original_mdt = type.getFactory()
-                .createMetaData(buf);
+        StorableMessageMetaData originalMessageMetadata = type.getFactory().createMetaData(buf);
+
+        byte[] underlying;
+        //TODO need to implement factory pattern here
+        if (type.equals(MessageMetaDataType.META_DATA_MQTT)) {
+            underlying = MQTTMetaDataHandler.constructMetadata(routingKey, buf, originalMessageMetadata, exchangeName);
+        } else {
+            underlying = AMQPMetaDataHandler.constructMetadata(routingKey, buf, originalMessageMetadata, exchangeName);
+        }
+
+        return underlying;
+    }
+
+    /**
+     * Create a copy of metadata
+     *
+     * @param originalMetadata          source metadata that needs to be copied
+     * @param newCompressedMessageValue Value to indicate if the message is compressed or not
+     * @return copy of the metadata as a byte array
+     */
+    private byte[] createNewMetadata(byte[] originalMetadata, boolean newCompressedMessageValue) {
```

Comment:
Shall we change the name 'newCompressedMessageValue' to a more simple and readable value like compressed?